### PR TITLE
P2P GetTransactions

### PR DESCRIPTION
**Pull request by kirugan**
Issue #1074

### CODE REVIEW

File: p2p/starknet/handlers.go [hunk 1]
```diff
@@ -101,75 +101,118 @@ func (h *Handler) reqHandler(req *spec.Request) (Stream[proto.Message], error) {
...
+
+		return &spec.TransactionsResponse{
+			BlockNumber: block.Number,
+			BlockHash:   core2p2p.AdaptHash(block.Hash),
+			Responses: &spec.TransactionsResponse_Transactions{
+				Transactions: &spec.Transactions{
+					Items: utils.Map(block.Transactions, core2p2p.AdaptTransaction),
+				},
+			},
+		}, false
```
Comment (joshklop) [inline p2p/starknet/handlers.go:202]: Iterator implementation looks good to me overall.

Should we return `true` here, and return `false` only when `it.Err() != nil` or `!it.Valid()`?

File: p2p/starknet/iterator.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	return true
+}
+
+func (it *iterator) Next() bool {
+	if it.forward {
+		it.blockNumber += it.step
+	} else {
+		it.blockNumber -= it.step
+	}
+	it.limit--
```
Comment (joshklop) [inline p2p/starknet/iterator.go:56]: Can `limit` underflow?

File: p2p/starknet/iterator.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+	return it.Valid()
+}
+
+func (it *iterator) Block() *core.Block {
```
Comment (joshklop) [inline p2p/starknet/iterator.go:61]: Would it be better to return the error directly, rather than requiring the caller to query `it.Err()`?

File: p2p/starknet/iterator.go (hunk 2 above)
Comment (kirugan) [reply p2p/starknet/iterator.go:56]: nice catch, thanks!

File: p2p/starknet/iterator.go (hunk 3 above)
Comment (kirugan) [reply p2p/starknet/iterator.go:61]: Yeah, I was thinking about this for a while. For some time it was a common pattern for iterators to have `Next()`, `HasNext()`, and `Err` (see https://pkg.go.dev/bufio#Scanner for example). I'm not sure about it now, will reconsider it

File: adapters/core2p2p/transaction.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		case tx.Version.Is(1):
+			specTx.Txn = &spec.Transaction_InvokeV1_{
+				InvokeV1: &spec.Transaction_InvokeV1{
+					Sender:    AdaptFeltToAddress(tx.SenderAddress),
+					MaxFee:    AdaptFelt(tx.MaxFee),
+					Signature: AdaptSignature(tx.Signature()),
+					Calldata:  AdaptFeltSlice(tx.CallData),
+				},
+			}
+		case tx.Version.Is(3):
```
Comment (omerfirmak) [inline adapters/core2p2p/transaction.go:132]: Let's not do version 3s for now. They will not be added until 0.13.0

File: adapters/core2p2p/transaction.go
Comment (IronGauntlets) [inline adapters/core2p2p/transaction.go:1]: I think it would be useful to have some tests for this. As in the `starknet_test.go` we only check that we have received the correct number of transactions

**Review by IronGauntlets — CHANGES_REQUESTED** [review]

File: p2p/starknet/starknet_test.go [hunk 5]
```diff
@@ -179,15 +180,138 @@ func TestClientHandler(t *testing.T) {
 	})
 
 	t.Run("get txns", func(t *testing.T) {
-		res, cErr := client.RequestTransactions(testCtx, &spec.TransactionsRequest{})
+		blocks := []*core.Block{
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply p2p/starknet/starknet_test.go:183]: If `core2p2p.AdaptTransaction` is covered %100, it should be fine

File: adapters/core2p2p/transaction.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+				DeployAccountV1: &spec.Transaction_DeployAccountV1{
+					MaxFee:      AdaptFelt(tx.MaxFee),
+					Signature:   AdaptAccountSignature(tx.Signature()),
+					ClassHash:   AdaptHash(tx.ClassHash),
+					Nonce:       AdaptFelt(tx.Nonce),
+					AddressSalt: AdaptFelt(tx.ContractAddressSalt),
+					Calldata:    AdaptFeltSlice(tx.ConstructorCallData),
+				},
+			}
+		case tx.Version.Is(3):
```
Comment (omerfirmak) [inline adapters/core2p2p/transaction.go:35]: We can omit V3 txns for now I believe. But feel free to ignore that suggestion.

File: adapters/core2p2p/transaction.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+		Deploy: &spec.Transaction_Deploy{
+			ClassHash:   AdaptHash(tx.ClassHash),
+			AddressSalt: AdaptFelt(tx.ContractAddressSalt),
+			Calldata:    AdaptFeltSlice(tx.ConstructorCallData),
+		},
+	}
+}
+
+func adaptL1HandlerTransaction(tx *core.L1HandlerTransaction) *spec.Transaction_L1Handler {
+	if tx.Version != nil && !tx.Version.Is(1) {
```
Comment (omerfirmak) [inline adapters/core2p2p/transaction.go:153]: shouldnt this be;

```suggestion
	if tx.Version == nil || !tx.Version.Is(1) {
```

Comment (omerfirmak) [reply adapters/core2p2p/transaction.go:153]: We can assume it is non-nil btw

Comment (IronGauntlets) [reply adapters/core2p2p/transaction.go:119]: Yeah because all of the transactions from the feeder have versions

File: p2p/starknet/starknet_test.go [hunk 8]
```diff
@@ -270,13 +350,103 @@ func TestClientHandler(t *testing.T) {
...
+	var f felt.Felt
+	f.SetUint64(v)
 
-	return sl
+	txV := core.TransactionVersion(f)
+	return &txV
+}
+
+func feltSlice(n int) []*felt.Felt {
+	return make([]*felt.Felt, n)
```
Comment (IronGauntlets) [inline p2p/starknet/starknet_test.go:449]: nit: consider populating the slice with rand felts

File: p2p/starknet/starknet_test.go [hunk 9]
```diff
@@ -287,3 +457,52 @@ func randFelt(t *testing.T) *felt.Felt {
 
 	return f
 }
+
+func fillFelts[T any](t *testing.T, i T) T {
```
Comment (IronGauntlets) [inline p2p/starknet/starknet_test.go:461]: I think this should be moved to `utils` package along with `randFelt` and `feltSlice` as they can be useful in other tests. However, we can do it in another pr.

File: p2p/starknet/starknet_test.go [hunk 10]
```diff
@@ -179,15 +174,100 @@ func TestClientHandler(t *testing.T) {
...
+				Step:      1,
+			},
+		})
 		require.NoError(t, cErr)
 
-		count := uint64(0)
+		var count uint64
 		for txn, valid := res(); valid; txn, valid = res() {
+			if count == numOfBlocks {
+				assert.NotNil(t, txn.GetFin())
```
Comment (IronGauntlets) [inline p2p/starknet/starknet_test.go:260]: I think we need to add a testcase to make sure when multiple blocks's transactions are requested then the fin has been sent correctly. However, that can be done in your other pr.

**Review by IronGauntlets — APPROVED** [review]
Left some suggestions for future PR. Otherwise, everything looks good.

File: adapters/core2p2p/transaction.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+					Calldata:    AdaptFeltSlice(tx.ConstructorCallData),
+				},
+			}
+		default:
+			panic(fmt.Errorf("unsupported InvokeTransaction version %s", tx.Version))
+		}
+	case *core.DeclareTransaction:
+		switch {
+		case tx.Version == nil:
+			panic("Declare transaction has not set version field")
```
Comment (IronGauntlets) [inline adapters/core2p2p/transaction.go:41]: These should be removed as well.
